text/0000-libobs-keychain-api.md

```diff
@@ -0,0 +1,78 @@
+# Summary
+
+This RFC proposes the addition of OS keychain APIs to the libobs platform utilities to faciliate secure storage of credentials.
+
+# Motivation
+
+While the credentials stored by OBS in profiles are not of severe sensitivity due to limited scope, as a modern desktop application we should still do our best to store them securely.
+
+# Detailed Changes
+
+This RFC proposes the addition of the following libobs API methods:
+
+- `bool os_keychain_available(void)`
+- `bool os_keychain_save(const char *label, const char *key, const char *data)`
+- `bool os_keychain_load(const char *label, const char *key, char **data)`
+- `bool os_keychain_delete(const char *label, const char *key)`
+
+As the API requires OS-specific implementations it shall become part of the `util/platform.h` header.
+
+**Implementation notes:**
+
+Operating-system backends:
+- Windows: `wincred.h` / `Cred*` Win32 API functions
+- macOS: Keychain Services API
+- Linux: `libsecret`
+
+The `label` is a user-visible component that may be displayed in keychain management applications as the label or name of an entry. It must be specified, ASCII only, and must be human-readable.  
+In most cases it should identify the component and contents, e.g. for OAuth credentials stored by the OBS Studio UI this could be "OBS Studio OAuth Credentials".  
+An entry can only be loaded or deleted if the label and key match the ones specified when saving.
+
+`os_keychain_available()` will always return `true` on Windows and macOS, on Linux it will return `true` if libobs was compiled with libsecret support, but does not indicate whether a keychain backend is available.
+
+**Usage examples:**
+Saving:
+```c
+const char *label = "OBS Studio Stream Keys";
+const char *key = "StreamKey_abcef";
+const char *value = "mystreamkey";
+bool success = os_keychain_save(label, key, value);
+```
+
+Loading:
+```c
+const char *label = "OBS Studio Stream Keys";
+const char *key = "StreamKey_abcef";
+char *value = NULL;
+bool success = os_keychain_load(label, key, &value);
+
+... do stuff ...
+
+if (value)
+    bfree(value);
+```
+
+Deletion:
+```c
+const char *label = "OBS Studio Stream Keys";
+const char *key = "StreamKey_abcef";
+bool success = os_keychain_delete(label, key);
+```
+
+# Drawbacks
+
+Accessing keychain information can reduce portability of profiles by not storing information in config files. OBS Studio may opt out of using a keychain if it is running in portable mode.
```

Comment:
On macOS (and maybe Windows 11?) keychain items can be cloud-based, so using the keychain would _actually_ increase portability vs the status quo.

Reply:
The intention of portable mode is to have a self-contained instance that can be kept on e.g. a USB thumb drive I don't think that using a keychain makes sense there. And to avoid synchronisation issues in general I think the data should be kept local either to a machine or portable instance.

As far as I can tell credentials stored using the Win32 APIs on Windows aren't synced either, and macOS doesn't support portable mode.

Reply:
Right "portable" as in "stored with the application data", not "portable" as in "data follows me between machines".

Keep in mind that iCloud Keychain is the suggested "default" keychain on macOS - keeping application local is considered "deprecated".

Reply:
The current implementation in PR obsproject/obs-studio#9122 just uses the default/iCloud keychain.